File: KominioAi/backend/src/main/kotlin/com/kominioai/global/common/Result.kt
package com.kominioai.global.common

sealed class Result<out T> {
    data class Success<T>(val data: T) : Result<T>()
    data class Failure(val error: Throwable) : Result<Nothing>()
    
    fun <R> map(transform: (T) -> R): Result<R> = when (this) {
        is Success -> Success(transform(data))
        is Failure -> this
    }
    
    fun <R> flatMap(transform: (T) -> Result<R>): Result<R> = when (this) {
        is Success -> transform(data)
        is Failure -> this
    }
    
    fun onSuccess(action: (T) -> Unit): Result<T> {
        if (this is Success) action(data)
        return this
    }
    
    fun onFailure(action: (Throwable) -> Unit): Result<T> {
        if (this is Failure) action(error)
        return this
    }
} 

File: skkk/frontend/src/domain/Result.ts
import type { AppError } from "./errors";

export type Result<T, E = AppError> =
    | { success: true; value: T }
    | { success: false; error: E };

export function ok<T, E = AppError>(value: T): Result<T, E> {
    return { success: true, value };
}

export function err<T, E = AppError>(error: E): Result<T, E> {
    return { success: false, error };
}

export function map<T, U, E>(r: Result<T, E>, f: (t: T) => U): Result<U, E> {
    return r.success ? ok(f(r.value)) : r;
}

export function flatMap<T, U, E>(
    r: Result<T, E>,
    f: (t: T) => Result<U, E>
): Result<U, E> {
    return r.success ? f(r.value) : r;
}

export function fold<T, E, R>(
    r: Result<T, E>,
    onSuccess: (t: T) => R,
    onFailure: (e: E) => R
): R {
    return r.success ? onSuccess(r.value) : onFailure(r.error);
}

export function getOrElse<T, E>(r: Result<T, E>, defaultValue: T): T {
    return r.success ? r.value : defaultValue;
}
